File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
// check for the condition in which the LL has 0 or 1 element
        if(head == null || head.next == null){
            return false;
        }

        ListNode node1 = head;
        ListNode node2 = head;
        while(node2 != null && node2.next != null){
            node1 = node1.next; //move one step 
            node2= node2.next.next; //move two step 
            if(node1 == node2){
                return true;
            }
        }


    return false;


    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {

    // let nodesSet = new Set();
    // let curr = head;


    // //base case 
    // if(!head) return false

    // while(curr){

    //      if(nodesSet.has(curr)){
    //         return true;
    //      }else{
    //         nodesSet.add(curr);
    //         curr=curr.next;
    //      }





    // }


    // return false;

    // approach - 2

   

    if(head == null || head.next == null) return false;

     let slow = head;
    let fast = head.next;


    while(slow != fast){


       if(fast == null || fast.next == null) return false;

        slow = slow.next;
        fast=fast.next.next
    }


    return true;
};